Feature new streaming-viewer

BenjaminDeleze opened the pull request:
**Description**
This PR introduces the new streaming-viewer based on WREN.

**Tasks**
Functionalities
- [x] be able to display the same objects/worlds as the current version but with almost the same graphics as in Webots
- [x] support for animation
- [x] finish last todos in the code
- [x] Test on Mac
- [x] Capsule primitive
- [x] improve the Makefile to make the new streaming-viewer
- [x] support for other coordinate systems
- [x] Installation instruction for Linux
- [x] Installation instruction for MacOS
- [x] Installation instruction for Windows
- [ ] support for the documentation viewer
- [x] add support for different fonts (in PR #2976 )

Web interface
- [x] Improve the disconnect / reconnect
- [x] Add loading gif in animation
- [x] Reset the viewpoint when resetting the simulation 
- [ ] improve the navigation (rotation) in the viewer for mobile phone
- [x] improve the animation viewer with fullscreen mode and better looking playbar (in PR #2979 )

Cleanup
- [x] initial cleanup
- [x] Restore the ros submodule (it was deleted because it was creating issues with githubpages)
- [x] Remove generated files (needed for the use with github pages now)
- [x] Restore the files that have modifications unrelated to this PR
- [x] Cleanup the old code of Webotsjs
  - [x] Replace ```var``` by ```const``` and ```let```
  - [x] Delete unused files
  - [x] Get rid of the external dependency to load .hdr file
  - [x] Replace ```XMLHttpRequest``` with the ```fetch``` API when possible
- [x] Cleanup of the new code
  - [x] Cleanup of utils/
  - [x] Cleanup of nodes/
  - [x] Cleanup of wren/
- [x] Change animation template to point towards the right urls
- [x] Change index.html to point towards remote links
- [x] Add necessary components to webots server

Know Bugs
  - [x] Restore mjpeg streaming
  - [x] Viewpoint not reset correctly in animation with camera following an object
  - [ ] Viewpoint follow problem with animation
  - [ ] GTAO is lighter than the one of Webots due to glClipControl
  - [ ] Slowdown on Windows 10 and macOS with Firefox. Due to glLinkProgram on phongShader, phongStencilAmbientEmissiveShader, phongStencilDiffuseSpecularShader, pbrShader and pbrStencilDiffuseSpecularShader

omichel reviewed (commented):
Here is a first series of comments for my early review.
I also found that it takes quite some time to finalize the 3D scene after it's downloaded and before it's rendered, not sure what takes so much time...

BenjaminDeleze commented on .nojekyll:
```diff
@@ -0,0 +1 @@
+
```
It is to prevent github pages to build with jekyll

BenjaminDeleze replied:
Removed in 38b1924

BenjaminDeleze commented on resources/web/streaming_viewer/enum.js:
```diff
@@ -0,0 +1,91 @@
+const ENUM = {
```
Yes, so it should not be commited right?

omichel replied:
Yes, we should not commit auto-generated files.

omichel replied:
The generation of these files should be part of some Makefile.

BenjaminDeleze replied:
Removed in 7809b59

BenjaminDeleze commented on .gitmodules:
```diff
@@ -12,4 +12,4 @@
 [submodule "resources/webots_ros"]
 	path = resources/webots_ros
-	url = https://github.com/cyberbotics/webots_ros
+	url = https://github.com/cyberbotics/webots_ros/
```
reverted in b5cc492

BenjaminDeleze commented on index.html:
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
```
It was a test for github pages. Removed in 
38b1924

BenjaminDeleze commented on src/wren/functionsToExport.txt:
```diff
@@ -0,0 +1 @@
+_wr_shader_program_new, _wr_shader_program_delete, _wr_shader_program_set_vertex_shader_path, _wr_shader_program_set_fragment_shader_path, _wr_shader_program_use_uniform, _wr_shader_program_use_uniform_buffer, _wr_shader_program_create_custom_uniform, _wr_shader_program_set_custom_uniform_value, _wr_shader_program_get_gl_name, _wr_shader_program_has_vertex_shader_compilation_failed, _wr_shader_pr ...
```
Removed in 7809b59

BenjaminDeleze commented on resources/web/wwi/hdrLoader.js:
```diff
@@ -0,0 +1,322 @@
+ * @readonly
+ * @enum {number}
+ */
+export var WEBGL_PIXEL_FORMAT = {
```
fixed in 544a143

BenjaminDeleze commented on resources/web/wwi/file_loader.js:
```diff
@@ -0,0 +1,180 @@
+		// Check for data: URI
+		var dataUriRegex = /^data:(.*?)(;base64)?,(.*)$/;
+		var dataUriRegexResult = url.match(dataUriRegex);
+		var request;
```
fixed in 544a143

BenjaminDeleze commented on resources/web/wwi/file_loader.js:
```diff
@@ -0,0 +1,180 @@
+ * @param {zen3d.LoadingManager} manager — The loadingManager for the loader to use. Default is zen3d.DefaultLoadingManager.
+ */
+function FileLoader(manager) {
+	this.path = undefined;
```
fixed in 544a143

BenjaminDeleze commented on resources/web/streaming_viewer/libtess.min.js:
```diff
@@ -0,0 +1,58 @@
+/*
```
Fix in ab56d09

BenjaminDeleze commented on resources/web/wwi/file_loader.js:
```diff
@@ -0,0 +1,180 @@
+				}, 0);
+			}
+		} else {
+			request = new XMLHttpRequest();
```
fixed 86418a6

omichel commented:
Tests are failing on macOS, could you re-sync with the develop branch to fix it?

omichel commented:
The problem was indeed elsewhere:
```
src/wren/JSHelper.cpp:94:26: error: Signed integer overflow for expression 'd*d'. [integerOverflow]
    data[4 * d + 2] = (d * d) % 0xff;
```

BenjaminDeleze commented:
Yes, I corrected it but it is strange all the tests were passing when it was in draft

lukicdarkoo commented on resources/web/wwi/animation.js:
```diff
@@ -110,14 +122,14 @@ class Animation { // eslint-disable-line no-unused-vars
 
-  _updateAnimationState(requestedStep = undefined) {
-    var automaticMove = typeof requestedStep === 'undefined';
+  async _updateAnimationState(requestedStep = undefined) {
```
It should not be async as it doesn't call async function?

BenjaminDeleze replied:
fixed in 42914ba

lukicdarkoo commented on include/wren/JSHelper.h:
```diff
@@ -0,0 +1,36 @@
+// Copyright 1996-2021 Cyberbotics Ltd.
```
This is a C file, so it should be lowercase

BenjaminDeleze replied:
Thanks, fixed in 42914ba

lukicdarkoo commented on resources/web/wwi/nodes/wbIndexedFaceSet.js:
```diff
@@ -0,0 +1,28 @@
+import {WbTriangleMeshGeometry} from './wbTriangleMeshGeometry.js';
+
+class WbIndexedFaceSet extends WbTriangleMeshGeometry {
```
```suggestion
export default class WbIndexedFaceSet extends WbTriangleMeshGeometry {
```

Single class files should use `export default`, so you don't need brackets when importing (for example, `import WbGeometry from './wbGeometry.js'`). Also, why the filenames are lowercase, it should be the same as the class names.

BenjaminDeleze replied:
The old WebotsJS files were following this logical : filenames lowercase, class capitalcase. But I think it is a good idea to change it.

BenjaminDeleze replied:
changed applied in 37b92a4

lukicdarkoo commented on resources/web/streaming_viewer/setup_viewer.js:
```diff
@@ -26,32 +25,43 @@ function init() {
   view.setTimeout(-1); // disable timeout that stops the simulation after a given time
   const streamingMode = modeSelect.options[modeSelect.selectedIndex].value;
+
+  view.onready = function() {
```
Maybe it is a good practice to use arrow functions when `this` of the new function is not needed.
```suggestion
  view.onready = _ => {
```

BenjaminDeleze replied:
thanks, fixed in 8d9e151

lukicdarkoo reviewed (approved):
Very nice!

I did a very quick check and I dropped a few general comments

stefaniapedrazzi commented:
Personally I think that in this PR there is now a big issue with file naming conventions.
We can use any coding style we want. But it should be consistent.

There are two main coding styles commonly used: lowercase + underscore or camel case.
Currently there are file names like `Default_url.js`, `Setup_viewer.js`, `Multimedia_client` and many others that doesn't follow any commonly used naming conventions.

I saw @lukicdarkoo asked to change the file name of classes to camel case to match the class name.
This could be fine. But this means to not only capitalize the first word of the file name, but to change the filename to camel case (so for example `multimedia_client.js` should be changed to `MultimediaClient.js` and not `Multimedia_client.js`).
Same for all the other files that doesn't contain any class: the filename should be changed to the camel case style (or keep the lowercase + underscore style) but for sure not be changed to the mixed capitalized first word + underscore + lowercase.

Note that the same issues with file names applies also to the other PRs based on this one.

BenjaminDeleze commented:
You are definitly right! I had a look at [Google JavaScript Style Guide](https://google.github.io/styleguide/jsguide.html#file-name). They defined the correct coding style as follows:
- File names must be all lowercase and may include underscores 
- Class, interface, record, and typedef names are written in UpperCamelCase
- Method names are written in lowerCamelCase
- Enum names are written in UpperCamelCase

This PR was following these set of rules (except for Enum, I have to change that) before the request of @lukicdarkoo.

But I kind of agree with him that it would make more sense to name the files in UpperCamelCase to match the name of the classes (without _ of course). I saw some other git repo that were doing so.

I saw in other places that people were using different naming conventions to match the content of the files:
- File that would contain a class: UpperCamelCase
- File that regroup functions : lowerCamelCase

But i am not very fond off this one.

omichel commented:
I agree with this.
For files that contains a class definition, the filename should match exactly the class name (as in C++).
For files that do not contain a class definition, but utility functions, the filename should be in `snake_case` (as in C).
This way we can understand just by seeing the file name whether it contains a class or not.
Moreover `snake_case` is typical for non-object oriented stuff and therefore clearly means that the file contains no class, but rather functions.

stefaniapedrazzi commented:
:eyes:

stefaniapedrazzi commented:
What should I review exactly? Is this branch fully functional and "ready" to be merged?

Because I recompiled Webots in this branch using the main Makefile, tried to record an animation and it doesn't work.
The first error messages are about `resources/web/streaming_viewer/enum.js` and `resources/web/streaming_viewer/wrenjs.js` not found.
If they are ignored and not generated or installed during the main compilation, where should they come from?

stefaniapedrazzi commented:
Meanwhile, I found some unrelated changes in this PR that should be reverted:
* `docs/discord/update.py`
* `tests/rendering/worlds/.mirror.wbproj`

BenjaminDeleze commented:
Yes the branch is normally fully functional.

The files you are missing are generated during the compilation of Webots if the prerequisites are installed. You can find the procedure to follow here : https://github.com/cyberbotics/webots/wiki/Linux-Optional-Dependencies. (Sorry I should have mentioned that before)

Thanks I will revert them

stefaniapedrazzi commented:
It seems there are also still some issues with the file names:
* `Webots.js` it is not a class so it seems should be renamed to `webots.js`
* ` include/wren/jsHelper.h` to be consistent with C naming conventions should be renamed to `js_helper.js`

BenjaminDeleze commented:
I realize that you will need additional informations to test it properly.

I will put it back to draft to modify what you suggested and write some additonal comments.

stefaniapedrazzi commented:
> The files you are missing are generated during the compilation of Webots if the prerequisites are installed. You can find the procedure to follow here : https://github.com/cyberbotics/webots/wiki/Linux-Optional-Dependencies. (Sorry I should have mentioned that before)

All the dependencies should be automatically installed.
Wou should also the required instructions in the `scripts/install/linux_optional_compilation_dependencies.sh`.

stefaniapedrazzi commented:
> I realize that you will need additional informations to test it properly.

If I need additional information other than what is provided in the documentation. Then it seems that this PR is not ready to be merged in the develop branch.
All what a user needs to know should be provided in the documentation.
If something is working differently than before, then the documentation should be updated.

BenjaminDeleze commented:
> All the dependencies should be automatically installed.

I think that we discussed with @omichel that this would be an optional dependency as the resources needed to play animations normally will be on the Cyberbotics server

stefaniapedrazzi commented:
> I think that we discussed with @omichel that this would be an optional dependency as the resources needed to play animations normally will be on the Cyberbotics server

So then the index should be changed to link to the online resources and the resources needed to be tested needs to be uploaded.
I can test locally for now. But we should in any case test again once all the resources are available online.

stefaniapedrazzi commented:
Why the `enum.js` and `wrenjs.js` (and others) are stored in the `resources/web/streaming_viewer` folder?
This seems a generic file required not only by the streaming viewer but also by the animation, etc. so at first sight the `resources/web/wwi` would be a better location.

BenjaminDeleze commented:
> * tests/rendering/worlds/.mirror.wbproj

This file is not supposed to be pushed on github right ? (as gitignore ignore .wbproj). Is it ok if I open a PR to remove it from Master instead of restoring the file in this branch?

> Why the `enum.js` and `wrenjs.js` (and others) are stored in the `resources/web/streaming_viewer` folder?

enum.js could easily be stored in wwi. I renember that they were some issues to put wrenjs in another place due to the automatic generation of the wrenjs.data and wrenjs.wasm files but I will have a look again

stefaniapedrazzi commented:
> This file is not supposed to be pushed on github right ? (as gitignore ignore .wbproj). Is it ok if I open a PR to remove it from Master instead of restoring the file in this branch?

This file is required to run the test. Please do not remove it.

stefaniapedrazzi commented:
I compiled the needed resources locally, but I still get errors when trying to run an animation on Linux/Chrome:
```
Fetch API cannot load file:///home/stefi/Desktop/textured_shapes_2_2.x3d. URL scheme "file" is not supported.
loadWorldFile @ X3dScene.js:100
```

Does it work for you with Chrome?

BenjaminDeleze commented:
Do you run it on a local server ? 
If it is not the case you should try because of the limitations: https://cyberbotics.com/doc/guide/web-scene#remarks-on-the-used-technologies-and-their-limitations 

(I need to uptate the docs, but the limitations concerning the file urls where already true)

stefaniapedrazzi commented:
> Do you run it on a local server ?
If it is not the case you should try because of the limitations: https://cyberbotics.com/doc/guide/web-scene#remarks-on-the-used-technologies-and-their-limitations

Yes, I'm temporary testing taking all the resources are taken from a local server. But the animation file created and open by Webots will not be on the local server.

BenjaminDeleze commented:
> But the animation file created and open by Webots will not be on the local server.
I am afraid that they will need to be. As written in the doc: ```Some web browsers (for example Chrome and Firefox 68 or later) cannot open local files using the file protocol by default, while this is required by the Webots player to open the X3D file and the textures.```

I just tested it on Master to be sure and it is the same thing. I did not modified how the x3d file is open.

stefaniapedrazzi commented:
I tested a little bit the "streaming".
The rendering and scene loading seems generally fine and also the scene interaction seems fine.

But I cannot run the simulation: when I press run in Webots it immediately will be changed back to pause mode by the web interface.
Moreover when I press start the viewpoint in the web interface is reset to the original one and should not.


Then, in some cases there are some errors in the console:
* `samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt` (and also in other world files):
    ```
    Uncaught RuntimeError: function signature mismatch
    at vd (<anonymous>:wasm-function[1064]:0x4a27f)
    at WbTransform.applyRotationToWren (http://localhost/wwi/nodes/WbTransform.js:20:5)
    ```
* `spot.wbt`: ```Background : Incomplete cubemap```


Robot windows messages are not handled correctly and throw errors.
If they are meant to be restore later, then we should open an issue about it to be addressed before releasing a new version containing this changes.

stefaniapedrazzi commented:
>  I did not modified how the x3d file is open.

Yes, the function to get the file has been changed. Now it uses `fetch` that seems to not support the local files.

stefaniapedrazzi commented:
I tested a little bit the "streaming".
The rendering and scene loading seems generally fine and also the scene interaction seems fine.

Here are some issues I had when streaming in x3d mode:
1) Connecting to Webots from the streaming_viewer while loading a world make Webots crashing
2) I cannot run the simulation: when I press run in Webots it immediately will be changed back to pause mode by the web interface.
3) when I press start the viewpoint in the web interface is reset to the original one and should not.
4) in some cases there are some errors in the console:
    *  spot.wbt: ```Background : Incomplete cubemap```
5) robot window messages are not supported. if it is planned to restore them later, then it would be better to open an issue that have to be addressed before releasing the web interface changes.

stefaniapedrazzi commented on resources/web/wwi/Animation.js:
```diff
@@ -1,27 +1,23 @@
-        this._setup(JSON.parse(xmlhttp.responseText));
-    };
-    xmlhttp.send();
+    fetch(this.url)
+      .then(response => response.json())
+      .then(data => this._setup(data))
+      .catch((error) => console.log(error));
```
Here you changed the way the URL is retrieved.
This could explain why local files can no longer be opened (at least on Chrome)

stefaniapedrazzi commented on resources/web/wwi/MouseEvents.js:
```diff
@@ -210,41 +317,86 @@ class MouseEvents { // eslint-disable-line no-unused-vars
-      var moveX2 = x1 - this.state.x1;
-      var moveY1 = y - this.state.y;
-      var moveY2 = y1 - this.state.y1;
-      var ratio = window.devicePixelRatio || 1;
```
We had issues with retina displays due to the `window.devicePixelRatio`.
If possible we should test this PR also on retina displays.

stefaniapedrazzi commented on resources/web/wwi/Selector.js:
```diff
@@ -0,0 +1,62 @@
+    Selector.preciseId = 'n' + id;
+  }
+
+  static checkIfParentisSelected(node) {
```
```suggestion
  static checkIfParentIsSelected(node) {
```

stefaniapedrazzi reviewed (changes requested):
Here are some other comments regarding the code

BenjaminDeleze commented:
Thank you for all your precious feedbacks. I will take care of it

stefaniapedrazzi commented:
I don't see any major issue.
So it would be good for the next review if it would be possible to already test with the online resources instead of running a local server.

BenjaminDeleze commented on src/wren/demo/default.cpp:
```diff
@@ -14,12 +14,13 @@
 /*
  * Create a fixed-size window using glut, initialize WREN,
- * and display a sphere using the Webots default shader (without colors and with a texture).
+ * and display several colored sphere using the Webots phong shader.
```
I restored it to the original default.cpp

BenjaminDeleze commented on src/wren/GlState.cpp:
```diff
@@ -119,6 +131,13 @@ namespace wren {
+#ifdef __EMSCRIPTEN__
+      int array[4];
+      array[0] = -1;
+      // glGetIntegerv(GL_TEXTURE_FREE_MEMORY_ATI, array);
```
fixed in 771ea89

BenjaminDeleze commented:
I think I addressed all the feedbacks:
- Move wrenjs.js, enum.js,... to wwi/
- Revert to xmlhttp instead of fetch to allow local playing of the animations
- Correct the bug that made Webots crash when connecting during loading.
- The viewpoint is not reset anymore when pressing play.
- Address the error messages (incomplete irradiance cubemap and functions mismatches)
- Ignore robot windows messages
- Correct the remarks on the coding style and the naming of files
- Change default coordinates system to ENU
- Test on emulated retina display
- Removed some unused files in wwi/

All the resources to play animations are now available online. (the links in the animation template and index.html are pointing to them)

You can find instructions on how to build and test them locally [here](https://github.com/cyberbotics/webots/wiki/Wrenjs)

stefaniapedrazzi commented:
:eyes:

stefaniapedrazzi reviewed (changes requested):
Thank you, all the previous comments seems to have been addressed.

Here are some other issues I found:
- Objects selection in X3D streaming mode /animation is sometimes wrong:
  for example in `samples/robotbenchmark/object_avoidance/worlds/object_avoidance.wbt`, (depending on the cube selected) it sometimes selects multiple cubes instead of just one as in Webots. I can provide more details if needed.
- when pressing the quit button during X3D streaming, the message "Bye bye.." never disappears and seems that something is still running (lady bug is moving)
- MJPEG streaming:
  - when connecting with the streaming_viewer in MJPEG mode the "Connect/Disconnect" button is disabled
  - when connecting with the streaming_viewer in MJPEG mode, the first image is not displayed automatically but you have to click on the page to see it.
  - when pressing the quit button during MJPEG streaming, there is an error message in the console `WebSocket is already in CLOSING or CLOSED state.` and is is never properly closed (but it is never closed also for x3d streaming)

General comments:
- previously the saving dialog (animation, model, etc.) was automatically adding an index to save filename to avoid overwriting the file but currently this mechanism has been removed. Suggesting a different filename seems better to not mix up JSON/HTML/X3D files from different save.
- fullscreen icon in animation toolbar is wrong but will probably be fixed in #2979

Using the online files:
- sometimes animation are loaded without the toolbar but I could not reproduce this systematically
- on Chrome animation works fine but only using a local server.
  The option to start it with `--allow-file-access-from-file` doesn't work. Documentation should be updated to use for example `--disable-web-security --user-data-dir="[some directory here]"`

Using local files:
- in the CORS script to open the animation files from outside the local server, this import line is missing ``from http.server import HTTPServer, SimpleHTTPRequestHandler, test`` and even after setting properly all the URL I get this error
  ```Failed to load module script: The server responded with a non-JavaScript MIME type of "". Strict MIME type checking is enforced for module scripts per HTML spec.```
- otherwise instructions are fine but it should also be specified where are located the `wwi.css`, `enum.js`, etc whose path have to be replaced in the HTML file

BenjaminDeleze commented:
I fixed most of the issues but I could not reproduce some of them:

> * when pressing the quit button during MJPEG streaming, there is an error message in the console `WebSocket is already in CLOSING or CLOSED state.` and is is never properly closed (but it is never closed also for x3d streaming)

I did not managed to reproduced that. Was it systematic ?

> * fullscreen icon in animation toolbar is wrong but will probably be fixed in #2979

Yes exactly


> * sometimes animation are loaded without the toolbar but I could not reproduce this systematically

I could not reproduce it, do you have examples of worlds or situtations? Was it on a specific browser?

> * on Chrome animation works fine but only using a local server.
>   The option to start it with `--allow-file-access-from-file` doesn't work. Documentation should be updated to use for example `--disable-web-security --user-data-dir="[some directory here]"`

Ok here it is strange. I tested to open animation (linked to resources on the web) with Chrome before your review and it was working. I retested on Thursday morning and it was failing. But Thursday afternoon and Friday morning it is working again.
In all the cases, the other solution you suggest: `--disable-web-security --user-data-dir="[some directory here]"` was working in any case. I guess I will update the documentation to support both.

> * in the CORS script to open the animation files from outside the local server, this import line is missing `from http.server import HTTPServer, SimpleHTTPRequestHandler, test` and even after setting properly all the URL I get this error
>   `Failed to load module script: The server responded with a non-JavaScript MIME type of "". Strict MIME type checking is enforced for module scripts per HTML spec.`

When you specified the URLs, did you set them as files URLs like `/web/wwi/init_animation.js` or like http URL: `http:localhost:8000/web/wwi/init_animation` ? In the first case it is normal that it does not work, but in the second case it should work. (I will write it more explicitly in the doc).

stefaniapedrazzi commented:
> When you specified the URLs, did you set them as files URLs like /web/wwi/init_animation.js or like http URL: http:localhost:8000/web/wwi/init_animation ? In the first case it is normal that it does not work, but in the second case it should work. (I will write it more explicitly in the doc).

I used the URL starting with "http:localhost:8000/".

stefaniapedrazzi commented:
>> when pressing the quit button during MJPEG streaming, there is an error message in the console WebSocket is already in CLOSING or CLOSED state. and is is never properly closed (but it is never closed also for x3d streaming)

>I did not managed to reproduced that. Was it systematic ?

Yes, it was systematical. But now I can no longer reproduce it too.

Just also note, that if the broadcast checkbox is set, the "Quit the simulation" is enabled but nothing happen when clicked. Probably it should be disabled (similarly to the other simulation buttons).

stefaniapedrazzi commented on src/webots/gui/WbMainWindow.cpp:
```diff
@@ -1617,12 +1616,28 @@ void WbMainWindow::exportHtml() {
-    assert(!fileName.isEmpty());
+  QString fileName;
+  for (int i = 0; i < 1000; ++i) {
+    QString suffix = i == 0 ? "" : QString("_%1").arg(i);
```
```suggestion
    const QString suffix = i == 0 ? "" : QString("_%1").arg(i);
```

stefaniapedrazzi commented on src/webots/gui/WbMainWindow.cpp:
```diff
@@ -1617,12 +1616,28 @@ void WbMainWindow::exportHtml() {
   WbSimulationState::Mode currentMode = WbSimulationState::instance()->mode();
   WbSimulationState::instance()->setMode(WbSimulationState::PAUSE);
   WbWorld *world = WbWorld::instance();
+  QString worldName = QFileInfo(world->fileName()).baseName();
```
```suggestion
  const QString worldName = QFileInfo(world->fileName()).baseName();
```

stefaniapedrazzi commented on docs/guide/web-scene.md:
```diff
@@ -81,7 +81,7 @@ Here are some workarounds:
 
 - disable browser security flags:
-    - Chrome: launch with the `--allow-file-access-from-files` option
+    - Chrome: launch with the `--allow-file-access-from-files` option or with the `--disable-web-security --user-data-dir="[some directory here]"`.
```
Seems better like this:
```suggestion
    - Chrome: launch with the `--allow-file-access-from-files`  or `--disable-web-security --user-data-dir="[some directory here]"` option.
```

stefaniapedrazzi reviewed (changes requested):
I reviewed the latest commits and they are fine.
But rather than duplicating the code to compute the HTML file name with suffix in the `WbMainWindow::exportHtml` and `WbMainWindow::startAnimationRecording` functions it would be better to move the code in a dedicated function.

I tested the X3D and MJPEG streaming to check the latest fixes and I noticed some minor issues:
* in case of X3D streaming, the simulation is paused when the client connects, but in case on MJPEG the simulation is not paused.
* there are errors connecting using first the X3D and then the MJPEG mode and viceversa using the same streaming_viewer page (not reloading the the page when restarting Webots to change mode). It seems that some initialization is not correctly cleaned. In case of MJPEG (first X3D and then MJPEG), disconnecting and reconnecting fixes the issue. But in the other case (first MJPEG and then X3D), the scene remains completely white.

I will review and test again once the branch about the animation enhancement is merged.

BenjaminDeleze commented:
> * in case of X3D streaming, the simulation is paused when the client connects, but in case on MJPEG the simulation is not paused.

It is a behavior inherited from the threejs version. I think it is because the loading time is much bigger with x3d, so we want to pause to not miss anything, whereas in mjpeg it is "not needed". But it could be a good time to harmonize the behaviour. So either :

- we unpause at the end of the loading of x3d
- Or we pause when connecting to mjpeg.
- Or we let it like it is

I don't know which one you prefer @stefaniapedrazzi @omichel

lukicdarkoo commented on resources/web/wwi/X3dScene.js:
```diff
@@ -0,0 +1,285 @@
+  loadWorldFile(url, onLoad) {
+    const prefix = this.prefix;
+    const renderer = this.renderer;
+    const xmlhttp = new XMLHttpRequest();
```
`fetch()`?

lukicdarkoo commented on resources/web/wwi/X3dScene.js:
```diff
@@ -0,0 +1,285 @@
+  }
+}
+
+export {X3dScene};
```
It is exported in the beginning, this looks unnecessary.

lukicdarkoo commented on resources/web/wwi/css/animation.css:
```diff
@@ -0,0 +1,315 @@
+ position: relative
+}
+
+.webotsLabel {
```
```suggestion
.webots-label {
```

BenjaminDeleze commented on resources/web/wwi/Animation.js:
```diff
@@ -0,0 +1,763 @@
+
+  init(onReady) {
+    this.onReady = onReady;
+    var xmlhttp = new XMLHttpRequest();
```
no, unfortunately fetch does not support loading local files, so we have to keeep xmlhttp

BenjaminDeleze commented on resources/web/templates/x3d_playback.html:
```diff
@@ -20,20 +20,28 @@
-      function init() {
-          var view = new webots.View(document.getElementById("view3d"));
-          view.open("%x3dFilename%");%setAnimation%
+      var Module = [];
```
No, because Module could be modified in Wrenjs.js

BenjaminDeleze commented on resources/web/wwi/Animation.js:
```diff
@@ -0,0 +1,763 @@
+    if (this.gui === 'real_time')
+      this._updateAnimationState();
+
+    window.requestAnimationFrame(() => { this._updateAnimation(); });
```
We need to keep the arrow function structure to get the right `this` context. It will not work otherwise

stefaniapedrazzi commented on resources/web/wwi/WebotsStreaming.js:
```diff
@@ -0,0 +1,52 @@
+  async _init() {
+    let promises = [];
+    promises.push(this._load('https://git.io/glm-js.min.js'));
+    promises.push(this._load('https://cyberbotics.com/wwi/Wrenjs/enum.js'));
+    promises.push(this._load('https://cyberbotics.com/wwi/Wrenjs/wrenjs.js'));
```
Did you update the instructions to run everything locally?
Because these URLs now need to be changed too.

Same for WebotsAnimation.js

BenjaminDeleze replied:
I am currently adapting animations and the streaming viewer in web component following Darko's demand. That is why the urls changed. I will update the instructions once I am done with it.

stefaniapedrazzi reviewed (changes requested):
I now checked the animation and the new style and icons are nice.
But there are some issues, the first two only when the follow object is active (tested with `wall_following`):
- when jumping to a previous position the viewpoint, very often the robot is not visible because of the position of the viewpoint. I tested with the `wall_following.wbt` simulation but I didn't check if it was alrady the same before.
- if shadows are disabled, when jumping to a previous position, a sort of ghost robot representation remains in the scene where the robot was previously located. This disappears (becomes always less visible) after each new click on the window.
- the animation toolbar and all the animation buttons and labels should be shown above the simulation labels. But currently it seems that only the progrss bar is shown above and the speedometer is shown below. Tested with `propeller.wbt`.

Also note that if the floor color is light (close to white), it is difficult to read the speedometer values.

If needed I can provide some screenshots.

BenjaminDeleze commented on src/webots/gui/WbMainWindow.cpp:
```diff
@@ -2221,14 +2246,13 @@ void WbMainWindow::setWorldLoadingStatus(const QString &status) {
   WbSimulationState::Mode currentMode = WbSimulationState::instance()->mode();
-  WbSimulationState::instance()->setMode(WbSimulationState::PAUSE);
-  const QString &worldFileName = WbWorld::instance()->fileName();
+  QString fileName = findHtmlFileName("Save Animation File");
```
It does not work because the `WbProject::computeBestPathForSaveAs` does not take a const. (I think that is also why the other `QString fileName` of `WbMainWindow` are not const either.

BenjaminDeleze commented:
> * if shadows are disabled, when jumping to a previous position, a sort of ghost robot representation remains in the scene where the robot was previously located. This disappears (becomes always less visible) after each new click on the window.

This is a problem due to the way the ambient occlusion is calculated: it need to render several times to adapt.
This issues is also present on Webots desktop as you can see on the picture, we can still the ghosts of the barrels event so I deleted them (one by one, so ghosts of the first deleted barrel is less visible).
The only solution to this problem would be to render several time each time we jump back or forth, but I am not sure it is really a good thing to do.
![spot](https://user-images.githubusercontent.com/25938827/117947531-91afd380-b310-11eb-8b34-1f8a530c2d30.png)

stefaniapedrazzi commented:
> The only solution to this problem would be to render several time each time we jump back or forth, but I am not sure it is really a good thing to do.

Personally I think that a wrong rendering is much worse than having to rendering twice when jumping back and forth.
It is triggered by a  manual action, so even if it takes some more time it is not an issue.

BenjaminDeleze commented:
@stefaniapedrazzi I think I addressed all the feedbacks (except two of Darko but they are about style and I don't have time to tackle them today) if you want to have a look.

stefaniapedrazzi reviewed (changes requested):
Thank you. I tested a little bit and now it seems to work fine.
But more tests should be performed because it seems that some worlds are triggering errors.
For example streaming the city.wbt and after rotating the viewpoint I get this error:
```
Uncaught TypeError: Cannot read property 'indexOf' of undefined
    at X3dScene.applyPoseToObject (X3dScene.js:166)
    at X3dScene.applyPose (X3dScene.js:152)
    at X3dScene.processServerMessage (X3dScene.js:239)
    at Stream.onSocketMessage (Stream.js:123)
    at WebSocket.socket.onmessage (Stream.js:18)
```


Switching from MJPEG to X3D streaming and viceversa now works correctly.
However in case MJPEG first and X3D later or X3D connect after quitting with the quit button, I still get an error in the console:
I could reproduce this with `wall_following.wbt` but not systematically:
```
X3dScene.js:134 Uncaught TypeError: Cannot read property 'nodes' of undefined
    at X3dScene.applyPose (X3dScene.js:134)
    at X3dScene.processServerMessage (X3dScene.js:239)
    at Stream.onSocketMessage (Stream.js:123)
    at WebSocket.socket.onmessage (Stream.js:18)
```


Some documentation for running animation on localhost still needs to be updates:
https://github.com/cyberbotics/webots/wiki/Wrenjs#with-everything-on-a-single-local-server
And I tried again but the instructions in the "With only the animation player on a local server" bypass the CORS doesn't work for me on Linux/Chrome.

We should not forget before releasing Webots R2021b to create a `https://cyberbotics.com/wwi/R2021b/` folder, update all the URL to link to the R2021b folder and update the `new_version.sh` script to automatically update the URL when the major version changes.

BenjaminDeleze commented on resources/web/wwi/Animation.js:
```diff
@@ -0,0 +1,763 @@
+    document.addEventListener('slider_input', (e) => { those._updateSlider(e); });
+    document.getElementById('play-bar').appendChild(timeSlider);
+    document.querySelector('my-slider').shadowRoot.getElementById('range').addEventListener('mousemove', _ => this._updateFloatingTimePosition(_));
+    document.querySelector('my-slider').shadowRoot.getElementById('range').addEventListener('mouseleave', _ => this._hideFloatingTimePosition(_));
```
I am not sure it is a good idea in this case, because:

- We want to create the toolbar once the world is ready and the .json loaded and not before.
- It would result in a lot of DOM elements in which we are not really interested to have them as attributes because we use them only once

BenjaminDeleze commented:
> And I tried again but the instructions in the "With only the animation player on a local server" bypass the CORS doesn't work for me on Linux/Chrome.

I tried again and it works for me on ubuntu with chrome (with --allow-file-access-from-files). So I don't really know what to do.

stefaniapedrazzi commented:
> I tried again and it works for me on ubuntu with chrome (with --allow-file-access-from-files). So I don't really know what to do.

I will check again on a fresh install.
I'm testing on Ubuntu 18.04 with Chrome Version 90.0.4430.93 (Official Build) (64-bit)

lukicdarkoo commented on resources/web/wwi/nodes/utils/WbMatrix4.js:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  inverse() {
+    const inv = [];
```
```suggestion
    const inv = new Array(16);
```
I am surprised it was working before. This should be slightly faster.

BenjaminDeleze replied:
I don't like this syntax too much because I find it error prone. `new Array(16)` will effectively create an array of 16 but `new Array('16')` will create an array of length 1 with 16 as element. But yes it should slightly improve the performance so I will apply it.

lukicdarkoo commented on resources/web/wwi/nodes/WbWorld.js:
```diff
@@ -0,0 +1,22 @@
+  constructor() {
+    this.hasFog = false;
+    this.coordinateSystem = 'ENU';
+    this.upVector = glm.vec3(0, 1, 0);
```
```suggestion
    this.upVector = glm.vec3(0, 0, 1);
```
Should it be like this?

stefaniapedrazzi commented:
I tested on a clean installation of Ubuntu 18.04 with the latest Chrome and still `--allow-file-access-from-files` doesn't work.
However we can keep the documentation as is given that we give a working alternative.

But what is the purpose of the section "With only the animation player on a local server" in the wiki if you still have to use `--allow-file-access-from-files` or other ways to bypass CORS?

lukicdarkoo reviewed (changes requested):
A few details

BenjaminDeleze commented:
> But what is the purpose of the section "With only the animation player on a local server" in the wiki if you still have to use `--allow-file-access-from-files` or other ways to bypass CORS?

It was to be able to test the local files as if they were on the server, but without having to put them on the ftp server. But I guess that it is useless as long as the local files are the same as the one on the ftp server.

stefaniapedrazzi commented on docs/guide/web-scene.md:
```diff
@@ -81,7 +82,7 @@ Here are some workarounds:
 
 - disable browser security flags:
-    - Chrome: launch with the `--allow-file-access-from-files` option
+    - Chrome: launch with the `--allow-file-access-from-files`  or `--disable-web-security --user-data-dir="[some directory here]"` option.
```
I won't use `--user-data-dir=.` because it will pollute the local folder.
Usually something like this is used `--user-data-dir=".chrome"`.

stefaniapedrazzi commented:
I just started testing some worlds, and till now all the errors are fixed.
But there are strange viewpoint jumps on some simulations that doesn't occur in Webots, for example:
* tinkerbots_demo.wbt
* khepera3_gripper.wbt

stefaniapedrazzi commented on resources/web/wwi/Parser.js:
```diff
@@ -0,0 +1,1157 @@
+        else
+          await this._parseChildren(node, parent);
+      } else {
+        console.log(scene);
```
Is this a debug print or is it wanted to print the whole scene in the Console?

stefaniapedrazzi commented:
The `Track` node texture animation is no longer working in the web interface. See for example the `pick_and_place.wbt` simulation. Previously this was working as can be seen on robotbenchmark.net.

stefaniapedrazzi commented:
I'm done with the review and the tests.
Apart from the regression with the Track texture animation and the Viewpoint jumps everything is fine for me.

BenjaminDeleze commented:
> Apart from the regression with the Track texture animation and the Viewpoint jumps everything is fine for me.

This is now fixed

lukicdarkoo reviewed (approved):
Looks good to me, thank you

stefaniapedrazzi commented on resources/web/wwi/X3dScene.js:
```diff
@@ -173,17 +173,29 @@ export default class X3dScene {
+            object.applyTranslationToWren();
+        } else {
+          object.translation = translation;
+          if (WbWorld.instance.readyForUpdates) {
```
Is not also better to check if `object` is an instance of WbTextureTransform just to avoid updating the appearance if other `translation` messages are received?

BenjaminDeleze commented:
I changed the urls to point to R2021b

stefaniapedrazzi reviewed (changes requested):
Thank you.
Now both issues are fixed but the viewpoint follow object functionality is broken in the animation.

BenjaminDeleze commented:
Fixed. Thank for the huge reviewing work you have done on this PR

BenjaminDeleze commented:
@stefaniapedrazzi if it works for you, is it possible to approve it such that we can merge it today

stefaniapedrazzi reviewed (approved):
Thank you.
I confirm that the viewpoint seems now to work correctly both in streaming and animation.
